packages/dd-trace/test/appsec/iast/security-controls/index.spec.js

```diff
@@ -0,0 +1,356 @@
+
+    beforeEach(() => {
+      controls = new Map()
+      parse = sinon.stub().returns(controls)
+      startChSubscribe = sinon.stub()
+      endChSubscribe = sinon.stub()
+
+      const channels = {
+        'dd-trace:moduleLoadStart': {
+          subscribe: startChSubscribe
+        },
+        'dd-trace:moduleLoadEnd': {
+          subscribe: endChSubscribe
+        }
+      }
+
+      securityControls = proxyquire('../../../../src/appsec/iast/security-controls', {
+        'dc-polyfill': {
+          channel: name => channels[name]
+        },
+        './parser': {
+          parse
+        }
+      })
+    })
+
+    afterEach(() => {
+      securityControls.disable()
+    })
+
+    it('should call parse and subscribe to moduleLoad channels', () => {
+      controls.set('sanitizer.js', {})
+
+      const securityControlsConfiguration = 'SANITIZER:CODE_INJECTION:sanitizer.js:sanitize'
+      securityControls.configure({ securityControlsConfiguration })
+
+      sinon.assert.calledWithExactly(parse, securityControlsConfiguration)
+
+      sinon.assert.calledOnce(startChSubscribe)
+      sinon.assert.calledOnce(endChSubscribe)
+    })
+
+    it('should call parse and not subscribe to moduleLoad channels', () => {
+      const securityControlsConfiguration = 'invalid_config'
+      securityControls.configure({ securityControlsConfiguration })
+
+      sinon.assert.calledWithExactly(parse, securityControlsConfiguration)
+
+      sinon.assert.notCalled(startChSubscribe)
+      sinon.assert.notCalled(endChSubscribe)
+    })
+  })
+
+  describe('hooks', () => {
+    beforeEach(() => {
+      addSecureMark = sinon.stub().callsFake((iastContext, input) => input)
+
+      iastContext = {}
+      const context = {}
+
+      securityControls = proxyquire('../../../../src/appsec/iast/security-controls', {
+        '../taint-tracking/operations': {
+          addSecureMark
+        },
+        '../../../../../datadog-core': {
+          storage: {
+            getStore: sinon.stub().returns(context)
+          }
+        }
+      })
+
+      saveIastContext(context, {}, iastContext)
+    })
+
+    afterEach(() => {
+      securityControls.disable()
+      sinon.restore()
+    })
+
+    function requireAndPublish (moduleName) {
+      const filename = require.resolve(moduleName)
+      let module = require(moduleName)
+
+      const payload = { filename, module }
+      moduleLoadEndChannel.publish(payload)
+      module = payload.module
+      return module
+    }
+
+    it('should hook a module only once', () => {
+      // eslint-disable-next-line no-multi-str
+      const conf = 'INPUT_VALIDATOR:COMMAND_INJECTION:packages/dd-trace/test/appsec/iast\
+      /security-controls/resources/custom_input_validator.js:validate'
+      securityControls.configure({ securityControlsConfiguration: conf })
+
+      requireAndPublish('./resources/custom_input_validator')
+
+      requireAndPublish('./resources/custom_input_validator')
```

Comment:
where are you testing that it is not hooking twice?

Reply:
oops